lib/common.js

```diff
@@ -327,43 +327,77 @@ function matchStringOrRegexp(target, pattern) {
   return pattern instanceof RegExp  ? str.match(pattern) : str === String(pattern);
 }
 
-// return [newKey, newValue]
-function formatQueryValue(key, value, options) {
+function formatQueryParams(queryParams) {
+  var formattedQueryParams = {}
+  for (var q in queryParams) {
+    if (!_.isUndefined(formattedQueryParams[q])) {
+      continue;
+    }
+    formattedQueryParams[q] = formatQueryValue(queryParams[q]);
+  }
+  return formattedQueryParams;
+}
+
+function decodeQueryParams(queryParams) {
+  var decodedQueryParams = {}
+  for (var q in queryParams) {
+    if (!_.isUndefined(decodedQueryParams[q])) {
+      continue;
+    }
+    decodedQueryParams[percentDecode(q)] = formatQueryValue(queryParams[q], percentDecode);
+  }
+  return decodedQueryParams;
+}
+
+function encodeQueryParams(queryParams) {
+  var encodedQueryParams = {}
+  for (var q in queryParams) {
+    if (!_.isUndefined(encodedQueryParams[q])) {
+      continue;
+    }
+    encodedQueryParams[percentEncode(q)] = formatQueryValue(queryParams[q], percentEncode);
+  }
+  return encodedQueryParams;
+}
+
+function formatQueryValue(value, stringFunction) {
   switch (true) {
-    case _.isNumber(value): // fall-though
+    case _.isNumber(value): // fall-through
     case _.isBoolean(value):
       value = value.toString();
       break;
-    case _.isUndefined(value): // fall-though
+    case _.isUndefined(value): // fall-through
     case _.isNull(value):
       value = '';
       break;
     case _.isString(value):
-      if(options.encodedQueryParams) {
-        value = percentDecode(value);
+      if(stringFunction) {
+        value = stringFunction(value);
       }
       break;
     case (value instanceof RegExp):
       break;
     case _.isArray(value):
       var tmpArray = new Array(value.length);
       for (var i = 0; i < value.length; ++i) {
-        tmpArray[i] = formatQueryValue(i, value[i], options)[1];
+        tmpArray[i] = formatQueryValue(value[i], stringFunction);
       }
       value = tmpArray;
       break;
     case _.isObject(value):
       var tmpObj = {};
       _.forOwn(value, function(subVal, subKey){
-        var subPair = formatQueryValue(subKey, subVal, options);
-        tmpObj[subPair[0]] = subPair[1];
+        var formattedSubKey = subKey;
+        if (stringFunction) {
+          formattedSubKey = stringFunction(formattedSubKey);
+        }
+        tmpObj[formattedSubKey] = formatQueryValue(subVal, stringFunction);
       });
       value = tmpObj;
       break;
   }
 
-  if (options.encodedQueryParams) key = percentDecode(key);
-  return [key, value];
```

Comment:
I think this was done this way to support multi-value query parameters.

Reply:
@ierceg maybe it was done this way for multi-value parameters at some point, but at the moment `formatQueryValue` was always using `key` as a key in the `queries` object, and so the behaviour is not changed AFAICT.